### ISSUE TITLE
port Dataset / Dataloader?

### ISSUE THREAD

**Issue by tycho01** [hasktorch/hasktorch#317]
Pytorch offers a [`torch.utils.data` module](https://pytorch.org/docs/stable/data.html), offering implementations for [`Dataset`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataset.py) / [`Dataloader`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py).
I'm not sure if these are deemed in-scope here, but it'd be nice to have such a standardized way of dealing with data as well.

**Comment by tycho01** [comment]
oh, arguably the only bit in that `Dataset` file that isn't already built into Haskell's `Data.List` by default is maybe that [`random_split`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataset.py#L263)?
fwiw I tried a Haskell thing for that [here](https://github.com/tycho01/synthesis/blob/32e5523dacb0a4563196eed9c674a2c934ae7f97/library/Synthesis/Utility.hs#L182) (based on [this](https://wiki.haskell.org/Random_shuffle#Purely_functional)), if for a split to 3.

`Dataloader` functionality mostly seems `shuffle` (see `fisherYates` above) and `sampler` (just iterate) / `batch_sampler` ([this](https://github.com/hasktorch/hasktorch/blob/9efa5eebf1187d28bee2d9c73ef1bb11b9dd7b26/examples/static-mnist/Common.hs#L122-L153)), or something like this:

```haskell
-- | split a list into batches of a given size (or less for the last batch)
batchList :: Int -> [a] -> [[a]]
batchList n xs = batchList' n [] xs
  where
    batchList' n batches xs = let
        (batch, rest) = splitAt n xs
        batches' = batches <> [batch]
      in case rest of
          [] -> batches'
          _  -> batchList' n batches' rest
```

Dunno why our `Torch.Vision` still felt the need for a custom `length` then over just like being a `List` but :woman_shrugging:.

Anyway guess their data module isn't looking so vital anymore. :sweat_smile:

**Comment by tycho01** [comment]
in retrospect, I I feel there actually is a non-trivial challenge here: type-safe batching, i.e. on the type-level make guarantees about a batch sampler returning a static batch size, while PyTorch's batch sampler's behavior is... *not* to actually return a fixed size given the last batch may have less?
so getting some kind of official solution for that would seem pretty nice.

**Comment by tycho01** [comment]
closing in favor of prior discussion at #267.
